P4TR: post about upgrading LN to take advantage of taproot's benefits (#632)
bitcoinops/bitcoinops.github.io: _includes/specials/taproot/en/10-ln-with-taproot.md

*By [ZmnSCPxj][]*

In this post, we'll look at two privacy features that [taproot][topic taproot] enables
for LN:

* [PTLCs][topic ptlc] over LN.
* P2TR Channels.

### PTLCs Over LN

PTLCs enable [many features][suredbits payment points], with a major
feature for LN being [payment decorrelation][p4tr ptlcs] without
any need to randomize routes.[^route-randomization] Every node along a
single-path or [multipath][topic multipath payments] route can be given a scalar that is used to
tweak each forwarded PTLC, enabling *payment decorrelation* where
individual forwards no longer leak the unique identifier for each
LN payment.

> PTLCs are ***not a privacy panacea***.  If a surveillor node sees a

Author: Will fix, and I'll add a commit to this PR putting it in the style guide since it appears it's not just me that wants to use "surveillor".

> forward with a particular timelock and value, and a second surveillor

Reviewer: ```suggestion
forward with a particular timelock and value, and a second surveillant
```

payment path, even if the surveillor nodes can no longer correlate them
via a unique identifying hash.  However, we *do* get:

* Increased uncertainty in the analysis.  The probabilities surveilors
  can work with are now lower and thus their information is that much
  less valuable.
* A *lot* more decorrelation in multipath payments.  Separate paths will
  not have strong timelock and value correlation with each other, and if
  LN succeeds, there should be enough payments that timing
  correlation is not reliable either.
* No increase in cost compared to an [HTLC][topic htlc] (and possibly even a slight
  cost reduction due to [multisignature efficiency][p4tr
  multisignatures]).

In principle, a pre-taproot channel can be upgraded to support PTLCs without
closing and reopening the channel.  Existing channels can host PTLCs by
creating an offchain transaction that spends the existing non-taproot
funding output to a taproot output containing a PTLC.  Thus getting
support for PTLCs over LN does not require any cost to users
beyond each node and its channel peers upgrading their software.

However, to actually use PTLCs, *every* forwarding node from the spender

> to the receiver must support PTLCs.  This means PTLC support may remain

Reviewer: I might have misremembered that. At least Rene writes that it's highly unlikely that one will be able to mix HTLCs and PTLCs along the route.

> to the receiver must support PTLCs.  This means PTLC support may remain

Author: FWIW, @zmnscpxj and I also discussed this when he first submitted his draft of this article, and that was also his conclusion (and I found his argument persuasive).

existence isn't gossiped on LN.

Unfortunately, every LN channel is a 2-of-2, and in the current
pre-taproot Bitcoin, every 2-of-2 is *openly* coded.  LN is the
most popular user of 2-of-2 multisignature, so any blockchain explorer
can see a 2-of-2 being spent and guess with fairly good probability that
this is a LN channel being closed.  The funds can then be traced
from there, and if it goes to another P2WSH output, then that is likely
to be *another* unpublished channel.  Thus, even unpublished channels
are identifiable onchain once they are closed, with some level of false
positives.

Taproot, by using [schnorr signatures][topic schnorr signatures], allows for n-of-n to look exactly
the same as 1-of-1.  With some work, even [k-of-n][topic threshold signature] will also look the same
as 1-of-1 (and n-of-n).  We can then propose a feature where an LN
channel is backed by a P2TR UTXO, i.e. a
P2TR channel, which increases the *onchain* privacy of
unpublished channels.[^two-to-tango]

<!-- P2WSH 2-of-2: OP_0 <sig> <sig> <2 <key> <key> 2 OP_CMS>
             219 =   1 + 1+72 +1+72 +1+1+1+33+1+33+1+1
             54.75 = 219/4
     P2TR: <sig>

>              64

Reviewer: ```suggestion
             66 = 1+1+64
```

>              16 = 64/4

Reviewer: ```suggestion
             16.5 = 66/4
```

> <!-- P2WSH 2-of-2: OP_0 <sig> <sig> <2 <key> <key> 2 OP_CMS>
>              219 =   1 + 1+72 +1+72 +1+1+1+33+1+33+1+1
>              54.75 = 219/4
>      P2TR: <sig>
>              64
>              16 = 64/4

Reviewer: ```suggestion
<!-- P2WSH 2-of-2: itemCount OP_0 <sig> <sig> <2 <key> <key> 2 OP_CMS>
             220 =   1 + 1 + 1+72 +1+72 +1+1+1+33+1+33+1+1
             55 = 220/4
     P2TR: itemCount <sig>
             66 = 1 + 1 + 64
             16.5 = 66/4
```

>     Comparsion:

Reviewer: It's not entirely clear to me what this comparison is meant to illustrate, but it seems to me that this is meant to sum the size of just the witness stacks for 2-of-2 P2WSH and P2TR. If so, the P2WSH calculation does count the bytes for item count and item lengths (and looks correct), while the P2TR one does not account for witness item count and item length, I think it should be `1+1+64=66`.

>     Comparsion:

Author: Just want to note for posterity here, that the math was added by me, not ZmnSCPxj, so all these mistakes and omissions were my fault.

      38.75 = 54.75 - 16

>       ~70% = 1 - 16/54.75

Reviewer: ```suggestion
      ~70% = 1 - 16.5/54.75
```

This only compares witness stacks, see below.

-->

This (rather small) privacy boost also helps published channels as well.
Published channels are only gossiped until they are closed, so somebody
trying to look for published channels will not be able to learn about

> *historical* channels.  If a surveillor wants to see every published

Reviewer: ```suggestion
*historical* channels.  If a surveillant wants to see every published
```

channel, it has to store all that data itself and cannot rely on any
kind of "archival" node.

> In addition, taproot keypath spends are 38.5 vbytes (70%) smaller than

Reviewer: ```suggestion
In addition, taproot keypath spends are 39.25 vbytes (40%) smaller than
```

> In addition, taproot keypath spends are 38.5 vbytes (70%) smaller than
> LN's existing P2WSH spends.  Unfortunately, you **cannot upgrade

Reviewer: ```suggestion
In addition, taproot keypath spends are 39.25 vbytes (40%) smaller than
LN's existing P2WSH spends.  Unfortunately, you **cannot upgrade
```

This only compares the witness stack, not the full input size. Counting the `prevout`, `scriptSig`, and `nSequence`, a P2TR keypath input is **57.5 vB**, and a 2-of-2 P2WSH input should be **96 vB** (95.5 vB assuming signature grinding). This comes out to a 40% cost reduction for inputs.

PTLCs over LN is [dual-funding][topic dual funding].  Lisa Neigut created an initial
proposal for a dual-funding protocol in [BOLTs #524][], with the [first
dual-funded channel on mainnet][neigut first dual funded] being
[opened][first dual funded tx] almost 2 years and 6 months later.
Dual-funding only requires compatibility with your direct peers.  PTLCs
over LN require compatibilty with all routing nodes on your
selected paths, including the receiver, so I feel justified in giving
this feature a +50% time modifier due to the added complication, for an
estimate of 3 years and 9 months starting from when a specific PTLC
protocol is proposed.

For P2TR channels, we should note that while this is "only"
between two direct peers, it also has lower benefits.  Thus, I expect it
will be lower priority.  Assuming most developers prioritize
PTLC-over-LN, then I expect P2TR channels will start
getting worked on by the time the underlying [SIGHASH_NOINPUT][topic sighash_noinput] or other
ways to implement Decker-Russell-Osuntokun ("[Eltoo][topic eltoo]") are available.

[^route-randomization]:
    A payer can choose a very twisty path (i.e. route randomization) to

>     make HTLC correlation analysis wrong, but that has its drawbacks:

Reviewer: ```suggestion
    make HTLC correlation analysis wrong, but that has its own drawbacks:
```

    * Twisty paths are costlier *and* less reliable (more nodes
      have to be paid, and more nodes need to *successfully* forward
      in order for the payment to reach the destination).
    * Twisty paths are longer, meaning the payer is telling *more*
      nodes about the payment, making it *more* likely they will hit

>       *some* surveillor node.

Reviewer: ```suggestion
      *some* surveillant node.
```
